url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/lexer/Consumer.java

```diff
@@ -0,0 +1,581 @@
+			}
+		}
+		
+		return len;
+	}
+
+
+
+	/** Return contained consumer when having only one and no constraints. This is called immediately after construction. */
+	Consumer optimize()	{
+		if (constraints == null && sequence.size() == 1 && sequence.get(0) instanceof Consumer)	{
+			// give up this formal-only consumer, take rule to sub-consumer
+			Consumer cc = (Consumer) sequence.get(0);
+			cc.rule = rule;
+			return cc;
+		}
+		return this;
+	}
+
+	/**
+		Returns true if the rule of this consumer matches the passed left recursive rule.
+		E.g. passing "nonterm ::= nonterm something" would match "nonterm ::= something".
+	*/
+	boolean matchesRepeatableRule(Rule rule)	{
+		if (rule.rightSize() != this.rule.rightSize() + 1)
+			return false;
+		for (int i = 0; i < this.rule.rightSize(); i++)
+			if (this.rule.getRightSymbol(i).equals(rule.getRightSymbol(i + 1)) == false)
+				return false;
+		return true;
+	}
+
+
+
+	// consume methods
+
+	/** Passes the factory for Strategy objects to all contained ConsumerAlternatives. */
+	public void setStrategyFactoryMethod(StrategyFactoryMethod strategyFactoryMethod)	{
+		for (int i = 0; i < sequence.size(); i++)	{
+			Object c = sequence.get(i);
+			if (c instanceof Consumer)	{
+				((Consumer) c).setStrategyFactoryMethod(strategyFactoryMethod);
+			}
+		}
+	}
+
+	/**
+		Reads from input. Returns null if input did not match, else a result tree containing the text.
+		Ensures that a nullable consumer never returns null and a repeatable consumer repeats.
+		@param input Input object where to read from.
+		@return null if no match, else scanned input as a result tree.
+	*/
+	public ResultTree consume(InputText input)
+		throws IOException
+	{
+		// prepare scanning for results
+		ResultTree result = isRepeatable() ? ensureResultTree(null) : null;	// prepare a list container when repeatable
+		ResultTree r;
+		Token.Address start = new Token.Address(input.getScanLine(), input.getScanColumn(), input.getScanOffset());
+
+		// consume input and optionally loop when having a repeatable rule
+		do	{
+			r = consumeInternal(input);
+			if (r != null && r.hasText())
+				result = (result == null) ? r : result.addChild(r);
+		}
+		while (r != null && isRepeatable());
+
+		// check the result tree element
+		if (result != null && isRepeatable() && result.getChildCount() <= 0)
+			result = null;
+			
+		if (result == null && isNullable())	// having read no input should not break loop of caller
+			result = ensureResultTree(null);	// return empty result element
+		
+		// add range when there was a result
+		if (result != null)	{
+			Token.Address end = new Token.Address(input.getScanLine(), input.getScanColumn(), input.getScanOffset());
+			result.setRange(new Token.Range(start, end));
+		}
+		
+		return result;
+	}
+
+
+	/**
+		Consumes characters from Input and stores it into returned result tree.
+		Returns null if nothing has been consumed.
+	*/
+	protected ResultTree consumeInternal(InputText input)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.